### PR TITLE
Normative: Arbitrary module namespace identifier names

### LINKED ISSUE #40594: Support “Arbitrary module namespace identifier names”
## Search Terms

- Arbitrary module namespace identifier names
- Arbitrary module namespace identifiers
- Arbitrary module identifier names
- Arbitrary module identifiers
- Arbitrary module identifier
- Arbitrary export names
- Arbitrary export identifiers
- string export names
- string exports

## Suggestion

Implement support for <https://github.com/tc39/ecma262/pull/2154>.

## Use Cases

This is necessary for **WASM** module integration.

This would also allow typing [`@types/webidl‑conversions`](https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/types/webidl-conversions) without needing to use `export =`.

---

For transpiled modules, this can already be supported, since it’s just creation of arbitrary properties on the **CommonJS** `exports` object, but native modules will require `module: "ESNext"`.

## Examples

See <https://github.com/tc39/ecma262/pull/2154>

```ts repro
// @showEmit
// @filename: module.d.ts
export const foo: unknown;

declare const some_imports: unknown;
export { some_imports as "some imports" };

declare const another_imports: unknown;
export { another_imports as "another imports" };

declare const extra_imports: unknown;
export { extra_imports as "extra imports" };

declare const rest_exports: unknown;
export { rest_exports as "rest exports" };

// @filename: index.ts
import { "some imports" as foo } from "./module.js";

export { foo as "some exports" } from "./module.js";
export { "another imports" as "another exports" } from "./module.js";
export { "rest exports" } from "./module.js";
export { "extra imports" as extra_imports } from "./module.js";

export * as "star exports" from "./module.js";
```

[Workbench Repro](https://www.typescriptlang.org/dev/bug-workbench/?#code/PTAEAEGcAsHsHcCiBbAlgFwLACgQQGaoA2ApgHYCGyJAXKMrACYCupAdI2+pDiQB4AHWACd0oAMawykMfliw6zMgGsyCMgG4cORiXFEKwkhKkzQkWNQD6qZENGRFKtfE29BIsQG9zlkjbtPSFAKYIAiC2pQW3tuMNAAXy1sHT0DIxNpMQo1dGgSYQDYx1AlVXVk-ljQHxzYPIKioJDwuobhaMCHeKTtbF19Q2NJLNB+dGEKJocnctdKj1Easb4JqZjm0NAw8cnO4p7k1MGMkbMjGSsqoNmXN2xrpZ8L9CvF7hbtl5WDxKPcMDgQikSjUOioMi6PhcHjYDZPbaRYzwuKfOSwRKgfDCSzbNjABgsdgAK0gYX+j28WPknwifh+QR6WJxyDxBKYrBIbFJ5Pc1R8YTa+Q6KLJtKFBQZ3Ux2NxYXxhM53LJC35XxIZkpYoSzLlCo5JJVfM8yx2qz2oviW126y6Hx1stZ8vZRK5PIp71AACpaTJDFLUY62YrDbzsEA)

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [x] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [x] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

## Related issues

- <https://github.com/microsoft/TypeScript/issues/43870>
- <https://github.com/microsoft/TypeScript/pull/49297>

### PR DESCRIPTION
**Pull request by Jack-Works**
Implementing https://github.com/tc39/ecma262/pull/2154/

### CODE REVIEW

**Comment by Jack-Works** [comment]
Made a full rewrite of this PR. Just get the compiler part to work. Will handle language service days later.

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -3415,7 +3420,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                         let suggestion: Symbol | undefined;
                         let suggestedLib: string | undefined;
                         // Report missing lib first
-                        if (nameArg) {
+                            if (nameArg) {
```
Comment (ExE-Boss) [inline src/compiler/checker.ts:3423]: **Nit:** errorneous whitespace change:
```suggestion
                        if (nameArg) {
```

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -4220,8 +4227,11 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
         }
     }
 
-    function reportNonExportedMember(node: Node, name: Identifier, declarationName: string, moduleSymbol: Symbol, moduleName: string): void {
-        const localSymbol = tryCast(moduleSymbol.valueDeclaration, canHaveLocals)?.locals?.get(name.escapedText);
+    function reportNonExportedMember(node: Node, name: ModuleExportName, declarationName: string, moduleSymbol: Symbol, moduleName: string): void {
+        // for import { "ident" as T } we can provide meaningful error message. (ident is defined locally)
+        // for import { "not ident" as T } we will still look for the locals of the target module symbol, but
+        // since it's impossible to define a local ident with invalid name, we will use the normal message "Module_0_has_no_exported_member_1"
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:4233]: This comment is a little bit distracting from the point of the function itself. The fact that the impossible cases just result in a missed lookup is not worth mentioning in my opinion, so I'd delete it.

**Comment by DanielRosenwasser** [comment]
I think the comment

> Default export symbol (technically not wholly internal, but included here for usability)

is the key here, and the same applies to `This`. those symbols are just for comparisons in various places. If anything relied on `this` or `default` being special entities within a symbol table, they'd probably already be broken.

Maybe @weswigham or @gabritto can back me up on that statement though.

**Pull request closed without merging**